# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [3]:
# This is given to us by Udacity, it is named: twitter_archive_enhanced.csv I have downloaded from 'project description'

# Importing our needed libraries
import pandas as pd
import numpy as np
import requests

In [4]:
# loading in the file
df_twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [5]:
# This is also given to us by Udacity, however we must use the `Requests` library to download it from a given URL.

# downloading files using requests
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)

open('image-predictions.tsv','wb').write(r.content)


335079

In [6]:
# load the tsv into a dataframe (making sure to seperate by tab)
df_predicted_images = pd.read_csv('image-predictions.tsv', sep = '\t')

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

### Querying the Twitter API
* Grabbing every tweet's **Retweet** and **Favorite** count using `Tweepy`

    Steps:
    
    * Use tweet IDs from the archive
    * Query the twitter API for each tweet's JSON
    * Store the data in a file called `tweet_json.txt`
    * Each JSON should be written to its own line
    * read the .txt file line by line into pandas DF storing:
        * Tweet ID
        * Retweet Count
        * Favorite Count
        
        
*note: the phone verificaiton has some trouble, so I went ahead and used the file provided by Udacity, I will double check the verification in the future.*

In [7]:
# # BIG NOTE: Oddly, this is also resulting in Fails instead of the real value, so I have uploaded the tweet-json.txt file 
# # which was included next to the code.

# import tweepy
# from tweepy import OAuthHandler
# import json
# from timeit import default_timer as timer

# # Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# # These are hidden to comply with Twitter's API terms and conditions
# consumer_key = 'HIDDEN'
# consumer_secret = 'HIDDEN'
# access_token = 'HIDDEN'
# access_secret = 'HIDDEN'

# auth = OAuthHandler(consumer_key, consumer_secret)
# auth.set_access_token(access_token, access_secret)

# api = tweepy.API(auth, wait_on_rate_limit=True)

# # NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# # df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# # change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# # NOTE TO REVIEWER: this student had mobile verification issues so the following
# # Twitter API code was sent to this student from a Udacity instructor
# # Tweet IDs for which to gather additional data via Twitter's API
# tweet_ids = df_twitter_archive.tweet_id.values
# len(tweet_ids)

# # Query Twitter's API for JSON data for each tweet ID in the Twitter archive
# count = 0
# fails_dict = {}
# start = timer()
# # Save each tweet's returned JSON as a new line in a .txt file
# with open('tweet_json.txt', 'w') as outfile:
#     # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
#     for tweet_id in tweet_ids:
#         count += 1
#         print(str(count) + ": " + str(tweet_id))
#         try:
#             tweet = api.get_status(tweet_id, tweet_mode='extended')
#             print("Success")
#             json.dump(tweet._json, outfile)
#             outfile.write('\n')
#         except tweepy.TweepError as e:
#             print("Fail")
#             fails_dict[tweet_id] = e
#             pass
# end = timer()
# print(end - start)
# print(fails_dict)

In [8]:
# link for this useful code: https://stackoverflow.com/questions/30088006/loading-a-file-with-more-than-one-line-of-json-into-pandas

df_tweets = pd.read_json('tweet-json.txt',lines=True)

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



I will be assessing every file we got above in the same order:
1. We Rate Dogs Archive
2. Tweet Image Predictions
3. Querying the Twitter API

Keys of assessment:
* Completeness
* Validity
* Accuracy
* Consistency 

at the end we will group the issues under Quality and Tidiness

### Asessing We Rate Dogs Archive

#### Visual Assesment
using pandas .head() and .sample()

In [9]:
# It's best to simply look at the head
df_twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [10]:
# It's also beneficial to look at a sample
df_twitter_archive.sample(15)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2079  670826280409919488                    NaN                  NaN   
1442  696754882863349760                    NaN                  NaN   
1040  744223424764059648                    NaN                  NaN   
1654  683449695444799489                    NaN                  NaN   
1317  706593038911545345                    NaN                  NaN   
2320  666437273139982337                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
1631  684481074559381504                    NaN                  NaN   
1114  732732193018155009                    NaN                  NaN   
1321  706310011488698368                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
85    876120275196170240                    NaN                  NaN   
432   820749716845686786                    NaN                  NaN   
918   756651752796094464                    NaN                  NaN   
436   820314633777061888                    NaN                  NaN   

                      timestamp  \
2079  2015-11-29 04:47:03 +0000   
1442  2016-02-08 17:58:03 +0000   
1040  2016-06-18 17:41:06 +0000   
1654  2016-01-03 00:47:59 +0000   
1317  2016-03-06 21:31:22 +0000   
2320  2015-11-17 02:06:42 +0000   
2327  2015-11-17 00:06:54 +0000   
1631  2016-01-05 21:06:19 +0000   
1114  2016-05-18 00:39:02 +0000   
1321  2016-03-06 02:46:44 +0000   
979   2016-07-04 15:00:45 +0000   
85    2017-06-17 16:52:05 +0000   
432   2017-01-15 21:49:15 +0000   
918   2016-07-23 00:46:50 +0000   
436   2017-01-14 17:00:24 +0000   

                                                 source  \
2079  <a href="http://twitter.com/download/iphone" r...   
1442  <a href="http://twitter.com/download/iphone" r...   
1040  <a href="http://twitter.com/download/iphone" r...   
1654  <a href="http://twitter.com/download/iphone" r...   
1317  <a href="http://twitter.com/download/iphone" r...   
2320  <a href="http://twitter.com/download/iphone" r...   
2327  <a href="http://twitter.com/download/iphone" r...   
1631  <a href="http://twitter.com/download/iphone" r...   
1114  <a href="http://twitter.com/download/iphone" r...   
1321  <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
85    <a href="http://twitter.com/download/iphone" r...   
432   <a href="http://twitter.com/download/iphone" r...   
918   <a href="http://twitter.com/download/iphone" r...   
436   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2079  Scary dog here. Too many legs. Extra tail. Not...                  NaN   
1442  This is Reptar. He specifically asked for his ...                  NaN   
1040  This is actually a pupper and I'd pet it so we...                  NaN   
1654  I just want to be friends with this dog. Appea...                  NaN   
1317  Here's a sneak peek of me on spring break. 10/...                  NaN   
2320  Here we see a lone northeastern Cumberbatch. H...                  NaN   
2327  This is a southern Vesuvius bumblegruff. Can d...                  NaN   
1631  Meet Pippa. She's an Elfin High Feta. Compact ...                  NaN   
1114  This is Ralpher. He's an East Guinean Flop Dog...                  NaN   
1321  Here's a very sleepy pupper. Thinks it's an ai...                  NaN   
979   This is Atticus. He's quite simply America af....                  NaN   
85    Meet Venti, a seemingly caffeinated puppoccino...                  NaN   
432   Meet Sunny. He can take down a polar bear in o...                  NaN   
918   This is Gert. He just wants you to be happy. 1...                  NaN   
436   We are proud to support @LoveYourMelon on thei...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2079           

#### Programmatic assessment

In [11]:
# Checking .info
df_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [12]:
# checking for duplicates on the tweet id
df_twitter_archive.tweet_id.duplicated().sum()

0

In [13]:
# checking for denominators that arent 10
np.sort(df_twitter_archive.rating_denominator.unique())

array([  0,   2,   7,  10,  11,  15,  16,  20,  40,  50,  70,  80,  90,
       110, 120, 130, 150, 170])

In [14]:
# checking the number of occurances
df_twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

**From what we can see, is that the number of occurances for these numbers are so low, and from the information given in the project details we can say that these might be inforrectly gathered so we have to manually check them.**

In [15]:
# we use this line to print the whole text:
pd.set_option("display.max_colwidth", -1)

# check the rows where the denomintor is not 10:
df_twitter_archive.query('rating_denominator != "10"')[['rating_numerator','rating_denominator']]

rating_numerator  rating_denominator
313   960               0                 
342   11                15                
433   84                70                
516   24                7                 
784   9                 11                
902   165               150               
1068  9                 11                
1120  204               170               
1165  4                 20                
1202  50                50                
1228  99                90                
1254  80                80                
1274  45                50                
1351  60                50                
1433  44                40                
1598  4                 20                
1634  143               130               
1635  121               110               
1662  7                 11                
1663  20                16                
1779  144               120               
1843  88                80                
2335  1                 2

In [16]:
# getting the text for the rows printed above: 
df_twitter_archive.query('rating_denominator != "10"').text

313     @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho                                                                       
342     @docmisterio account started on 11/15/15                                                                                                                                 
433     The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                      
516     Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx
784     RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…                             
902     Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                        

In [17]:
df_twitter_archive.query('rating_denominator != "10"').expanded_urls

313     NaN                                                                                                                                                                                                                                                            
342     NaN                                                                                                                                                                                                                                                            
433     https://twitter.com/dog_rates/status/820690176645140481/photo/1,https://twitter.com/dog_rates/status/820690176645140481/photo/1,https://twitter.com/dog_rates/status/820690176645140481/photo/1                                                                
516     https://www.gofundme.com/sams-smile,https://twitter.com/dog_rates/status/810984652412424192/photo/1                                                                                                     

### Asessing Tweet Image Predictions

#### Visual Assessment

In [18]:
df_predicted_images.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0  1        Welsh_springer_spaniel  0.465074  True    collie               
1  1        redbone                 0.506826  True    miniature_pinscher   
2  1        German_shepherd         0.596461  True    malinois             
3  1        Rhodesian_ridgeback     0.408143  True    redbone              
4  1        miniature_pinscher      0.560311  True    Rottweiler           

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665  True    Shetland_sheepdog    0.061428  True    
1  0.074192  True    Rhodesian_ridgeback  0.072010  True    
2  0.138584  True    bloodhound           0.116197  True    
3  0.360687  True    miniature_pinscher   0.222752  True    
4  0.243682  True    Doberman             0.154629  True

In [19]:
df_predicted_images.sample(15)

tweet_id  \
333   672160042234327040   
360   672640509974827008   
1487  782021823840026624   
44    666781792255496192   
1402  768970937022709760   
525   676606785097199616   
296   671362598324076544   
96    667724302356258817   
155   668815180734689280   
181   669328503091937280   
1655  810284430598270976   
1556  793210959003287553   
899   699788877217865730   
1262  748977405889503236   
1392  767122157629476866   

                                                                                      jpg_url  \
333   https://pbs.twimg.com/media/CVP9_beUEAAwURR.jpg                                           
360   https://pbs.twimg.com/media/CVWy9v-VAAALSoE.jpg                                           
1487  https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg                                           
44    https://pbs.twimg.com/media/CUDigRXXIAATI_H.jpg                                           
1402  https://pbs.twimg.com/ext_tw_video_thumb/768967618174877700/pu/img/4wfsrs0ZnQ5pstXm.jpg   
525   https://pbs.twimg.com/media/CWPKSGpWcAQN6mw.jpg                                           
296   https://pbs.twimg.com/media/CVEouDRXAAEe8mt.jpg                                           
96    https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg                                           
155   https://pbs.twimg.com/media/CUgb21RXIAAlff7.jpg                                           
181   https://pbs.twimg.com/media/CUnuuLEWEAAlKjN.jpg                                           
1655  https://pbs.twimg.com/media/Cz61ZD4W8AAcJEU.jpg                                           
1556  https://pbs.twimg.com/media/CwINKJeW8AYHVkn.jpg                                           
899   https://pbs.twimg.com/media/CbYmRHyWEAASNzm.jpg                                           
1262  https://pbs.twimg.com/media/CmTm-XQXEAAEyN6.jpg                                           
1392  https://pbs.twimg.com/media/CqVdiBJWIAEDZB4.jpg                                           

      img_num                           p1   p1_conf  p1_dog  \
333   1        pug                          0.561027  True     
360   1        Chesapeake_Bay_retriever     0.420155  True     
1487  1        golden_retriever             0.383223  True     
44    1        Italian_greyhound            0.618316  True     
1402  1        Pomeranian                   0.182358  True     
525   1        Loafer                       0.202999  False    
296   1        tub                          0.393616  False    
96    1        ibex                         0.619098  False    
155   1        redbone                      0.461172  True     
181   1        Siberian_husky               0.424202  True     
1655  1        malamute                     0.620768  True     
1556  1        doormat                      0.874431  False    
899   1        Border_terrier               0.355060  True     
1262  1        German_short-haired_pointer  0.742216  True     
1392  2        toy_poodle                   0.873841  True     

                     p2   p2_conf  p2_dog                        p3   p3_conf  \
333   French_bulldog     0.222114  True    Labrador_retriever        0.065456   
360   Cardigan           0.266030  True    Labrador_retriever        0.042514   
1487  cocker_spaniel     0.165930  True    Chesapeake_Bay_retriever  0.118199   
44    Weimaraner         0.151363  True    vizsla                    0.085989   
1402  golden_retriever   0.110658  True    mousetrap                 0.086399   
525   doormat            0.200411  False   malinois                  0.142300   
296   bathtub            0.383522  False   swimming_trunks           0.077301   
96    bighorn            0.125119  False   ram                       0.074673   
155   Italian_greyhound  0.270733  True    miniature_pinscher        0.109752   
181   Eskimo_dog         0.237660  True    malamute                  0.052572   
1655  Eskimo_dog         0.158395  True    Tibetan_mastiff           0.028962   
1556  French_bull

#### Programatic Assessment

In [20]:
df_predicted_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


**2075 predictions where we have 2356 tweets on the archive, therefore we must delete the incompatible ones from the archive**

In [21]:
df_predicted_images.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17  1.203855     0.594548     1.345886e-01  6.032417e-02
std    6.785203e+16  0.561875     0.271174     1.006657e-01  5.090593e-02
min    6.660209e+17  1.000000     0.044333     1.011300e-08  1.740170e-10
25%    6.764835e+17  1.000000     0.364412     5.388625e-02  1.622240e-02
50%    7.119988e+17  1.000000     0.588230     1.181810e-01  4.944380e-02
75%    7.932034e+17  1.000000     0.843855     1.955655e-01  9.180755e-02
max    8.924206e+17  4.000000     1.000000     4.880140e-01  2.734190e-01

In [22]:
# checking for duplicates on the tweet id
df_predicted_images.tweet_id.duplicated().sum()

0

In [23]:
# checking for duplicates on the jpg_url
df_predicted_images.jpg_url.duplicated().sum()

66

### Asessing Querying Tweets

#### Visual Assessment

In [24]:
df_tweets.head()

contributors  coordinates          created_at display_text_range  \
0 NaN           NaN          2017-08-01 16:23:56  [0, 85]             
1 NaN           NaN          2017-08-01 00:17:27  [0, 138]            
2 NaN           NaN          2017-07-31 00:18:03  [0, 121]            
3 NaN           NaN          2017-07-30 15:58:51  [0, 79]             
4 NaN           NaN          2017-07-29 16:00:24  [0, 138]            

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892420639486877696, 'id_str': '892420639486877696', 'indices': [86, 109], 'media_url': 'http://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg', 'url': 'https://t.co/MgUWQ76dJU', 'display_url': 'pic.twitter.com/MgUWQ76dJU', 'expanded_url': 'https://twitter.com/dog_rates/status/892420643555336193/photo/1', 'type': 'photo', 'sizes': {'large': {'w': 540, 'h': 528, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'small': {'w': 540, 'h': 528, 'resize': 'fit'}, 'medium': {'w': 540, 'h': 528, 'resize': 'fit'}}}]}                                               
1  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 892177413194625024, 'id_str': '892177413194625024', 'indices': [139, 162], 'media_url': 'http://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg', 'url': 'https://t.co/0Xxu71qeIV', 'display_url': 'pic.twitter.com/0Xxu71qeIV', 'expanded_url': 'https://twitter.com/dog_rates/status/892177421306343426/photo/1', 'type': 'photo', 'sizes': {'large': {'w': 1407, 'h': 1600, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'small': {'w': 598, 'h': 680, 'resize': 'fit'}, 'medium': {'w': 1055, 'h': 1200, 'resize': 'fit'}}}]}                                          
2  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 891815175371796480, 'id_str': '891815175371796480', 'indices': [122, 145], 'media_url': 'http://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg', 'url': 'https://t.co/wUnZnhtVJB', 'display_url': 'pic.twitter.com/wUnZnhtVJB', 'expanded_url': 'https://twitter.com/dog_rates/status/891815181378084864/photo/1', 'type': 'photo', 'sizes': {'medium': {'w': 901, 'h': 1200, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'large': {'w': 1201, 'h': 1600, 'resize': 'fit'}, 'small': {'w': 510, 'h': 680, 'resize': 'fit'}}}]}                                           
3  {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 891689552724799489, 'id_str': '891689552724799489', 'indices': [80, 103], 'media_url': 'http://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg', 'media_url_https': 'https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg', 'url': 'https://t.co/tD36da7qLQ', 'display_url': 'pic.twitter.com/tD36da7qLQ', 'expanded_url': 'https://twitter.com/dog_rates/status/891689557279858688/photo/1', 'type': 'photo', 'sizes': {'medium': {'w': 901, 'h': 1200, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'large': {'w': 1201, 'h': 1600, 'resize': 'fit'}, 'small': {'w': 510, 'h': 680, 'resize': 'fit'}

In [25]:
df_tweets.sample(15)

contributors  coordinates          created_at display_text_range  \
159  NaN           NaN          2017-05-06 18:27:40  [0, 82]             
349  NaN           NaN          2017-02-14 17:17:22  [0, 117]            
1294 NaN           NaN          2016-03-10 16:42:10  [0, 140]            
914  NaN           NaN          2016-07-23 23:42:53  [0, 102]            
2288 NaN           NaN          2015-11-19 02:43:18  [0, 75]             
1275 NaN           NaN          2016-03-13 15:43:18  [0, 140]            
6    NaN           NaN          2017-07-28 16:27:12  [0, 140]            
561  NaN           NaN          2016-11-26 19:50:26  [0, 100]            
1059 NaN           NaN          2016-06-10 16:19:48  [0, 115]            
1435 NaN           NaN          2016-02-10 02:14:42  [0, 130]            
2331 NaN           NaN          2015-11-16 19:31:45  [0, 139]            
1449 NaN           NaN          2016-02-06 02:22:53  [0, 102]            
963  NaN           NaN          2016-07-05 20:41:01  [0, 63]             
2196 NaN           NaN          2015-11-23 15:35:39  [0, 137]            
1842 NaN           NaN          2015-12-13 01:25:37  [0, 111]            

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             entities  \
159   {'hashtags': [], 'symbols': [], 'user_mentions': [{'screen_name': 'tallylott', 'name': 'tally', 'id': 363890752, 'id_str': '363890752', 'indices': [3, 13]}, {'screen_name': 'dog_rates', 'name': 'WeRateDogs™ (author)', 'id': 4196983835, 'id_str': '4196983835', 'indices': [48, 58]}], 'urls': [], 'media': [{'id': 860914470050295814, 'id_str': '860914470050295814', 'indices': [59, 82], 'media_url': 'http://pbs.twimg.com/media/C_KVJjFXcAYcQMI.jpg', 'media_url_https': 'https://pbs.twimg.com/media/C_KVJjFXcAYcQMI.jpg', 'url': 'https://t.co/6n8hzNihJ9', 'display_url': 'pic.twitter.com/6n8hzNihJ9', 'expanded_url': 'https://twitter.com/tallylott/status/860914485250469888/photo/1', 'type': 'photo', 'sizes': {'small': {'w': 510, 'h': 680, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 848, 'h': 1131, 'resize': 'fit'}, 'large': {'w': 848, 'h': 1131, 'resize': 'fit'}}, 'source_status_id': 860914485250469888, 'source_status_id_str': '860914485250469888', 'source_user_id': 363890752, 'source_user_id_str': '363890752'}]}   
349   {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 831552915512901632, 'id_str': '831552915512901632', 'indices': [118, 141], 'media_url': 'http://pbs.twimg.com/media/C4pE-I0WQAABveu.jpg', 'media_url_https': 'https://pbs.twimg.com/media/C4pE-I0WQAABveu.jpg', 'url': 'https://t.co/BCbmFYLrse', 'display_url': 'pic.twitter.com/BCbmFYLrse', 'expanded_url': 'https://twitter.com/dog_rates/status/831552930092285952/photo/1', 'type': 'photo', 'sizes': {'large': {'w': 1151, 'h': 1886, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'medium': {'w': 732, 'h': 1200, 'resize': 'fit'}, 'small': {'w': 415, 'h': 680, 'resize': 'fit'}}}]}       

#### Programatic Assessment

In [26]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2354 non-null datetime64[ns]
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null float64
id                               2354 non-null int64
id_str                           2354 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 n

In [27]:
df_tweets.describe()

contributors  coordinates  favorite_count  geo            id  \
count  0.0           0.0          2354.000000     0.0  2.354000e+03   
mean  NaN           NaN           8080.968564    NaN   7.426978e+17   
std   NaN           NaN           11814.771334   NaN   6.852812e+16   
min   NaN           NaN           0.000000       NaN   6.660209e+17   
25%   NaN           NaN           1415.000000    NaN   6.783975e+17   
50%   NaN           NaN           3603.500000    NaN   7.194596e+17   
75%   NaN           NaN           10122.250000   NaN   7.993058e+17   
max   NaN           NaN           132810.000000  NaN   8.924206e+17   

             id_str  in_reply_to_status_id  in_reply_to_status_id_str  \
count  2.354000e+03  7.800000e+01           7.800000e+01                
mean   7.426978e+17  7.455079e+17           7.455079e+17                
std    6.852812e+16  7.582492e+16           7.582492e+16                
min    6.660209e+17  6.658147e+17           6.658147e+17                
25%    6.783975e+17  6.757419e+17           6.757419e+17                
50%    7.194596e+17  7.038708e+17           7.038708e+17                
75%    7.993058e+17  8.257804e+17           8.257804e+17                
max    8.924206e+17  8.862664e+17           8.862664e+17                

       in_reply_to_user_id  in_reply_to_user_id_str  possibly_sensitive  \
count  7.800000e+01         7.800000e+01             2211.0               
mean   2.014171e+16         2.014171e+16             0.0                  
std    1.252797e+17         1.252797e+17             0.0                  
min    1.185634e+07         1.185634e+07             0.0                  
25%    3.086374e+08         3.086374e+08             0.0                  
50%    4.196984e+09         4.196984e+09             0.0                  
75%    4.196984e+09         4.196984e+09             0.0                  
max    8.405479e+17         8.405479e+17             0.0                  

       possibly_sensitive_appealable  quoted_status_id  quoted_status_id_str  \
count  2211.0                         2.900000e+01      2.900000e+01           
mean   0.0                            8.162686e+17      8.162686e+17           
std    0.0                            6.164161e+16      6.164161e+16           
min    0.0                            6.721083e+17      6.721083e+17           
25%    0.0                            7.888183e+17      7.888183e+17           
50%    0.0                            8.340867e+17      8.340867e+17           
75%    0.0                            8.664587e+17      8.664587e+17           
max    0.0                            8.860534e+17      8.860534e+17           

       retweet_count  
count  2354.000000    
mean   3164.797366    
std    5284.770364    
min    0.000000       
25%    624.500000     
50%    1473.500000    
75%    3652.000000    
max    79515.000000

### Quality issues

#### WeRateDogs Archive

1. (1)Type Change: tweet_id from int to string
2. (2)Type Change: timestamp, and retweed_status_timestamp from string to datetime

3. Some ratings have been gathered incorrectly (nominator and denominator) the indices and their problems:


| index| Incorrect | Correct |
| --- | --- | --- |
| 313 | 960 / 0 | 13 / 10 |  
| 342 | 11 / 15 | NOT A RATING |
| 516 | 24 / 7  | NOT A RATING |
| 784 | 9 / 11 | 14 / 10 |
| 1068| 9 / 11 | 14 /10 |
| 1165| 4 / 20 | 13 / 10 |
| 1202| 50 / 50 | 11 / 10 |
| 1662 | 7 / 11 | 10 / 10|
| 2335 | 1 / 2 | 9 / 10|

    - 342, 516, 1598  is not a rating tweet.
    - 784 is retweet of 1068

4. Regulate the nominator and denominator for easier assessment

5. Delete Retweets
6. Missing Expanded URLS 

#### Tweet Image Predictions

7. (3)Type Change: tweet_id from int to string

8. fix missing values 2075 rows instead of the 2356 in WeRateDogs Archive

9. drop duplicated images (66 duplicated)

#### Querying the Twitter API

10. (4)Type Change: id from int to string

11. Drop: contributors, coordinates, display_text_range, entities, extended_entities, favorited, geo, in_reply_to_screen_name, in_reply_to_status_id, in_reply_to_status_id_str, in_reply_to_user_id, in_reply_to_user_id_str, is_quote_status, lang, place, possibly_sensitive, possibly_sensitive_appealable, quoted_status, quoted_status_id, quoted_status_id_str, retweeted, retweeted_status, source, truncate

 

### Tidiness issues

1. doggo, floofer, pupper, and puppo are all under one categorical variable (one variable 4 columns)


2. Drop: in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, and retweeted_status_user_id from float to string
Low sample, too many missing values, won't be useful for analysis.


3. Create one categorical column for the categorical variable dog breed in image prediciton (and drop unnecessary columns which have been categorized) and other for prediction confidence


4. Merge all three cleaned DataFrames into one master DF

## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [28]:
# Make copies of original pieces of data
df_archive_copy = df_twitter_archive.copy()
df_images_copy = df_predicted_images.copy()
df_tweets_copy = df_tweets.copy()

### Cleaning WeRateDogs Archive

### Issue #1: Type Change: tweet_id from int to string

#### Define: using the `astype()` method we convert tweet_id from int to string. 

Doing this because we don't want to do any arithmetic operations on the ID.

#### Code

In [29]:
df_archive_copy['tweet_id'] = df_archive_copy['tweet_id'].astype(str)

#### Test

In [30]:
df_archive_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(2), o

In [33]:
# double checking the object type
type(df_archive_copy.tweet_id[0]) is str

True

### Issue #2:

#### Define

#### Code

#### Test

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:
1.

2.

3.

### Visualization